--- a/net/minecraft/core/registries/BuiltInRegistries.java
+++ b/net/minecraft/core/registries/BuiltInRegistries.java
@@ -350,7 +_,7 @@
 
          if (p_259410_ instanceof DefaultedRegistry) {
             ResourceLocation resourcelocation = ((DefaultedRegistry)p_259410_).m_122315_();
-            Validate.notNull(p_259410_.m_7745_(resourcelocation), "Missing default of DefaultedMappedRegistry: " + resourcelocation);
+            Validate.notNull(p_259410_.m_7745_(resourcelocation), "Missing default of DefaultedMappedRegistry: " + p_260209_.m_123023_() + ", Entry: " + resourcelocation);
          }
 
       });
